websockify/websockifyserver.py

```diff
@@ -700,6 +705,17 @@ def start_server(self):
 
         if self.listen_fd != None:
             lsock = socket.fromfd(self.listen_fd, socket.AF_INET, socket.SOCK_STREAM)
+        elif self.listen_sock != None:
+            # Make sure the socket does not already exist
+            try:
+                os.unlink(self.listen_sock)
+            except OSError:
+                if os.path.exists(self.listen_sock):
+                    raise
+            lsock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
+            lsock.bind(self.listen_sock)
+            os.chmod(self.listen_sock, self.listen_sock_mode)
```

Comment:
I wonder if this is race:y...

The safe way is to use umask. I.e.:

```python3
oldmask = os.umask(0777 ^ self.listen_sock_mode)
try:
    lsock.bind(self.listen_sock)
finally:
    os.umask(oldmask)
```

Reply:
Not 100% sure that there isn´t a race condition, so far it worked for me, but that does not say too much.

With ```chmod``` you`d be sure you´re only changing permissions on that one file, while with ```umask``` there might be ways you end up changing permissions on other files that are created simultaneously.

Also if there where a race condition, would not this suffer from the same issue? Just instead of ```chmod``` failing it would change back to the old umask before the file was created?

And, whats this for? ```0777 ^```

Reply:
> With chmod you`d be sure you´re only changing permissions on that one file (while being unfamiliar with the project and only looking at that one line of code), while with umask there might be ways you end up changing permissions on other files that are created simultaneously. Though, if you´re sure about that change can do ofc.

We don't have threads (at least not at this point), so the `finally` should be enough to make sure we don't mess with any other files.

> Also if there where a race condition, would not this suffer from the same issue? Just instead of chmod failing it would change back to the old umask before the file was created?

The race isn't about `chmod()` failing, but rather about someone opening the socket before `chmod()` has been run. This new code is safe since there is no point where the socket exists with the wrong permissions.

> And, whats this for? 0777 ^

`umask()` is inverted in that it states the bits you *don't want*, rather than the bits you *want*. Did you check if you go the expected result with the new code?

Reply:
Sorry for the long time without any response.

You´re right about that. It´s indeed needed.

I did make that change too. Besides the lack of tests it now looks good? Gonna have to take a look at the existing tests and see whats tested there. Seems hard to test proxies. It probably would be enough if I just manage to validate the new settings work and send any packet at all over the socket file. As all the rest should (tm) work the same.

Reply:
Accidentally posted this with the wrong account. Did not know I was logged in with another account. Mostly using gitlab myself. Should really delete the other one not much point in it.